File: backend/src/main/java/com/matriculaonline/service/TurmaService.java
package com.matriculaonline.service;

import com.matriculaonline.domain.exception.BusinessException;
import com.matriculaonline.domain.exception.DuplicateResourceException;
import com.matriculaonline.domain.exception.ResourceNotFoundException;
import com.matriculaonline.domain.model.Disciplina;
import com.matriculaonline.domain.model.StatusTurma;
import com.matriculaonline.domain.model.Turma;
import com.matriculaonline.dto.request.TurmaRequest;
import com.matriculaonline.dto.response.PageResponse;
import com.matriculaonline.dto.response.TurmaResponse;
import com.matriculaonline.repository.DisciplinaRepository;
import com.matriculaonline.repository.MatriculaRepository;
import com.matriculaonline.repository.TurmaRepository;
import com.matriculaonline.repository.TurmaSpecifications;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
public class TurmaService {

    private final TurmaRepository turmaRepository;
    private final DisciplinaRepository disciplinaRepository;
    private final MatriculaRepository matriculaRepository;

    public TurmaService(TurmaRepository turmaRepository,
                        DisciplinaRepository disciplinaRepository,
                        MatriculaRepository matriculaRepository) {
        this.turmaRepository = turmaRepository;
        this.disciplinaRepository = disciplinaRepository;
        this.matriculaRepository = matriculaRepository;
    }

    @Transactional
    public TurmaResponse criar(TurmaRequest request) {
        if (turmaRepository.existsByCodigo(request.codigo())) {
            throw new DuplicateResourceException("Já existe uma turma com o código: " + request.codigo());
        }

        Disciplina disciplina = disciplinaRepository.findByUuid(request.disciplinaUuid())
                .orElseThrow(() -> new ResourceNotFoundException("Disciplina", request.disciplinaUuid().toString()));

        Turma turma = new Turma();
        turma.setCodigo(request.codigo());
        turma.setDisciplina(disciplina);
        turma.setProfessor(request.professor());
        turma.setSemestre(request.semestre());
        turma.setVagas(request.vagas());
        turma.setStatus(request.status());
        return TurmaResponse.fromEntity(turmaRepository.save(turma));
    }

    @Transactional(readOnly = true)
    public PageResponse<TurmaResponse> listar(StatusTurma status, Boolean lotada, String q, Pageable pageable) {
        return PageResponse.from(
                turmaRepository.findAll(TurmaSpecifications.comFiltros(status, lotada, q), pageable),
                TurmaResponse::fromEntity
        );
    }

    @Transactional(readOnly = true)
    public TurmaResponse buscarPorUuid(UUID uuid) {
        Turma turma = turmaRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Turma", uuid.toString()));
        return TurmaResponse.fromEntity(turma);
    }

    @Transactional
    public TurmaResponse atualizar(UUID uuid, TurmaRequest request) {
        Turma turma = turmaRepository.findByUuid(uuid)
                .orElseThrow(() -> new ResourceNotFoundException("Turma", uuid.toString()));

        Disciplina disciplina = disciplinaRepository.findByUuid(request.disciplinaUuid())
                .orElseThrow(() -> new ResourceNotFoundException("Disciplina", request.disciplinaUuid().toString()));

        turma.setCodigo(request.codigo());
        turma.setDisciplina(disciplina);
        turma.setProfessor(request.professor());
        turma.setSemestre(request.semestre());
        turma.setVagas(request.vagas());
        turma.setStatus(request.status());
        return TurmaResponse.fromEntity(turmaRepository.save(turma));
    }

    /**
     * Remove a turma. Idempotente: se o recurso já não existir, retorna sem erro (no-op).
     */
    @Transactional
    public void deletar(UUID uuid) {
        turmaRepository.findByUuid(uuid).ifPresent(turma -> {
            if (matriculaRepository.existsByTurmaUuid(uuid)) {
                throw new BusinessException("Turma possui matrículas vinculadas.");
            }
            turmaRepository.delete(turma);
        });
    }
}

File: frontend/src/app/services/turma.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Turma, TurmaRequest } from '../models/turma.model';
import { PageResponse } from '../models/page.model';

export interface TurmaListFilters {
  status?: string;
  lotada?: boolean;
  q?: string;
}

@Injectable({ providedIn: 'root' })
export class TurmaService {
  private readonly apiUrl = '/api/turmas';

  constructor(private http: HttpClient) {}

  listar(
    page = 0,
    size = 10,
    sort = 'codigo,asc',
    filters: TurmaListFilters = {}
  ): Observable<PageResponse<Turma>> {
    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort);

    if (filters.status) {
      params = params.set('status', filters.status);
    }
    if (filters.lotada) {
      params = params.set('lotada', 'true');
    }
    if (filters.q) {
      params = params.set('q', filters.q);
    }

    return this.http.get<PageResponse<Turma>>(this.apiUrl, { params });
  }

  buscarPorUuid(uuid: string): Observable<Turma> {
    return this.http.get<Turma>(`${this.apiUrl}/${uuid}`);
  }

  criar(request: TurmaRequest): Observable<Turma> {
    return this.http.post<Turma>(this.apiUrl, request);
  }

  atualizar(uuid: string, request: TurmaRequest): Observable<Turma> {
    return this.http.put<Turma>(`${this.apiUrl}/${uuid}`, request);
  }

  deletar(uuid: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${uuid}`);
  }
}
